### PR TITLE
Use structs to avoid confusing C pointer wrappers

### PR DESCRIPTION
**Pull request by krlmlr**
When working on ADBC, a bit of time was lost because those pointers were used incorrectly. This PR uses pointers to anonymous `struct`s instead of plain `void` pointers, avoiding the following error kinds:

- `duckdb_database` vs. `duckdb_connection`
- `duckdb_database` vs. `duckdb_database*`

If we agree that this (or a similar approach) is useful, I can proceed with securing the four remaining typedefs which currently break the build.

### CODE REVIEW

**Comment by Mytherin** [comment]
Thanks for the PR! I think this is a good idea, as long as it can be done in a portable manner that does not break any of the builds.

**Comment by krlmlr** [comment]
Tests seem much better now, the failure in the extension test seems to be an internet outage.

**Comment by Tishj** [comment]
Yea that failure is fixed by a recent PR that changes pulling openssl from github instead of the old source

File: src/include/duckdb.h [hunk 1]
```diff
@@ -228,15 +228,37 @@ typedef struct {
...
+	void *__arrs;
+} * duckdb_arrow_schema;
+typedef struct _duckdb_arrow_array {
+	void *__arra;
+} * duckdb_arrow_array;
+// typedef struct { void *__lglt; } *duckdb_logical_type;
+// typedef struct { void *__dtck; } *duckdb_data_chunk;
+// typedef struct { void *__vctr; } *duckdb_vector;
+// typedef struct { void *__val; } *duckdb_value;
 typedef void *duckdb_logical_type;
```
Comment (Mytherin) [inline src/include/duckdb.h:262]: Is there a reason we are not wrapping these types as well?

**Review by Mytherin — COMMENTED** [review]
Thanks for fixing the tests! Looks good to me now, one minor question:

**Pull request merged**